src/devices/cpu/m6809/6x09dasm.cpp

```diff
@@ -49,50 +49,107 @@ u32 m6x09_base_disassembler::opcode_alignment() const
 	return 1;
 }
 
+//-------------------------------------------------
+//  constructor
+//-------------------------------------------------
+
+m6x09_base_disassembler::m6x09_base_disassembler(const opcodeinfo *opcodes, size_t opcode_count, uint32_t level)
+	: m_level(level)
+{
+	// create filtered opcode table
+	for (int i=0; i<opcode_count; i++)
+	{
+		if (opcodes[i].level() & level)
+		{
+			m_opcodes.insert(opcodes[i]);
+		}
+	}
+}
+
+//-------------------------------------------------
+//  transparent comparator support
+//-------------------------------------------------
+
+bool m6x09_base_disassembler::opcodeinfo::compare::operator()(opcodeinfo const& lhs, opcodeinfo const& rhs) const
+{
+	return lhs.opcode() < rhs.opcode();
+}
+
+bool m6x09_base_disassembler::opcodeinfo::compare::operator()(uint16_t opcode, opcodeinfo const& rhs) const
+{
+	return opcode < rhs.opcode();
+}
+
+bool m6x09_base_disassembler::opcodeinfo::compare::operator()(opcodeinfo const& lhs, uint16_t opcode) const
+{
+	return lhs.opcode() < opcode;
+}
+
 //-------------------------------------------------
 //  fetch_opcode
 //-------------------------------------------------
 
 const m6x09_base_disassembler::opcodeinfo *m6x09_base_disassembler::fetch_opcode(const data_buffer &opcodes, offs_t &p)
 {
+	int infloop = 0;
 	uint16_t page = 0;
 	const opcodeinfo *op = nullptr;
+
 	while(!op)
 	{
 		// retrieve the opcode
 		uint16_t opcode = page | opcodes.r8(p++);
 
 		// perform the lookup
-		auto iter = std::find_if(
-			m_opcodes.begin(),
-			m_opcodes.end(),
-			[opcode](const opcodeinfo &info) { return info.opcode() == opcode; });
+		auto iter = m_opcodes.find(opcode);
 
 		// did we find something?
 		if (iter == m_opcodes.end())
-			return nullptr;
-
-		// was this a $10 or $11 page?
-		switch (iter->mode())
 		{
-		case PG1:
-			page = 0x1000;
-			break;
+			// on the 6809 an unimplemented page 2 or 3 opcodes fall through to the first page.
+			if (!(m_level & HD6309_EXCLUSIVE) && (page != 0))
+			{
+				// backup the opcode pointer and disassemble the bare opcode.
+				--p;
+				page = 0;
+			}
+			else
+				// nothing to disassemble.
+				return nullptr;
+		}
+		else
+		{
+			// was this a $10 or $11 page?
+			switch (iter->mode())
+			{
+			case PG2:
+			case PG3:
+				if (page == 0)
+				{
+					// remember the page that comes first
+					page = iter->opcode() << 8;
+				}
+				else if (m_level & HD6309_EXCLUSIVE)
+				{
+					// multiple pages are illegal on the 6309
+					return nullptr;
+				}
+
+				if (infloop++ > 65534)
+				{
+					// detect infinite loop of page opcodes
+					return nullptr;
+				}
```

Comment:
As far as I can tell, this still isn’t going to clearly show the presence of the prefixes that shouldn’t be there.  I still think it’s better to disassemble it as a sequence of `.db` directives or something followed by the actual instruction with the final prefix even if the CPU gets into a state where it won’t execute that way.  Folding multiple prefixes like this is going to produce output that’s difficult to use when you try to use it on something that isn’t entirely filled with valid code.

Reply:
Implemented @cuavas 's suggestion